### PR TITLE
memo-ts: Make everything like token/js

### LINKED ISSUE #3454: Memo TS lib exports CJS only, needs to export CJS and ESM
I'm trying to use the `@solana/spl-token` package on a NextJS application and when I try to run this:

```
import { TOKEN_PROGRAM_ID } from '@solana/spl-token'
const tokenAccounts = await connection.getParsedTokenAccountsByOwner(publicKey, { programId: TOKEN_PROGRAM_ID });
```

I get this error:
```
./node_modules/@solana/spl-token/lib/esm/extensions/extensionType.mjs:5:0
Module not found: Can't resolve './memoTransfer.mjs'

Import trace for requested module:
./node_modules/@solana/spl-token/lib/esm/extensions/index.mjs
./node_modules/@solana/spl-token/lib/esm/index.mjs
./pages/index.tsx
```

But if instead run this:

```
      const tokenAccounts = await connection.getParsedTokenAccountsByOwner(publicKey, { programId: new PublicKey('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA') });
```

Everything works as expected.

Of course I'm using this second way for the time being, so it's not a gating issue, but thought I'd bring it up as it might be breaking something somewhere else?

### PR DESCRIPTION
**Pull request by joncinque**
#### Problem

The memo ts build is not compatible with cjs, since it only builds in esm.

#### Solution

Let's just make everything look exactly like token/js, including build steps, tsconfig files, output locations,

Fixes #3454

### CODE REVIEW

File: memo/ts/package.json [hunk 1]
```diff
@@ -1,18 +1,26 @@
...
+  "exports": {
+    "import": "./lib/esm/index.mjs",
+    "require": "./lib/cjs/index.js"
+  },
   "scripts": {
-    "build": "npm run clean && tsc",
-    "clean": "rimraf ./dist",
+    "build": "yarn clean && tsc -p tsconfig.json && tsc-esm -p tsconfig.json && tsc -p tsconfig.cjs.json",
+    "postbuild": "echo '{\"type\":\"commonjs\"}' > lib/cjs/package.json && echo '{\"type\":\"module\"}' > lib/esm/package.json",
+    "clean": "rimraf ./lib",
```
Comment (jordansexton) [inline memo/ts/package.json:21]: `rimraf` isn't listed as a devDependency, so we should add that.

File: memo/ts/package.json [hunk 2]
```diff
@@ -1,18 +1,26 @@
...
-  "main": "dist/index.js",
-  "types": "dist/index.d.ts",
+  "main": "lib/cjs/index.js",
+  "module": "lib/esm/index.mjs",
+  "types": "lib/types/index.d.ts",
   "type": "module",
   "sideEffects": false,
+  "exports": {
+    "import": "./lib/esm/index.mjs",
+    "require": "./lib/cjs/index.js"
```
Comment (jordansexton) [inline memo/ts/package.json:16]:
```suggestion
    "require": "./lib/cjs/index.js",
    "types": "./lib/types/index.d.ts"
```

We haven't done this in many other places yet, but adding this extra `types` path helps some IDEs and build tools.

File: memo/ts/README.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
+# `@solana/spl-token`
```
Comment (jordansexton) [inline memo/ts/README.md:1]:
```suggestion
# `@solana/spl-memo`
```

**Review by jordansexton — COMMENTED** [review]
Looks good! 

I'm thinking we should also move the path to `memo/js` similar to the Token version?

And the license in the package.json is listed as MIT, but we use Apache2 elsewhere, and there's no LICENSE file here.

File: memo/ts/package.json (hunk 1 above)
Comment (joncinque) [reply memo/ts/package.json:21]: Weird, I thought it was provided by default somewhere :sweat_smile:

File: memo/ts/README.md (hunk 3 above)
Comment (joncinque) [reply memo/ts/README.md:1]: Oops

**Comment by joncinque** [comment]
Should be good for another look

File: memo/js/README.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
+# `@solana/spl-memo`
+
+A TypeScript library for interacting with the SPL Memo program.
+
+## Links
+
+- [TypeScript Docs](https://solana-labs.github.io/solana-program-library/memo/ts/)
```
Comment (jordansexton) [inline memo/js/README.md:7]:
```suggestion
- [TypeScript Docs](https://solana-labs.github.io/solana-program-library/memo/js/)
```

File: memo/js/README.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+## Build from Source
+
+1. Clone the project:
+```shell
+git clone https://github.com/solana-labs/solana-program-library.git
+```
+
+2. Navigate to the library:
+```shell
+cd solana-program-library/memo/ts
```
Comment (jordansexton) [inline memo/js/README.md:26]:
```suggestion
cd solana-program-library/memo/js
```

**Comment by jordansexton** [comment]
Does this publish automatically or manually? https://github.com/solana-labs/solana-program-library/blob/10a70ddb3c769771d292c56ac04f74ca59acf488/token/js/package.json#L49 will need to update

**Comment by joncinque** [comment]
Thanks for the fixes and the merge.

> Does this publish automatically or manually?

No, we need to deploy this new version, then update token, then publish token.  Are you volunteering yourself to do all that? :eyes:

**Comment by jordansexton** [comment]
Sure, no problem!

**Pull request merged**